semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -134,48 +106,66 @@ let' n v m = do
   --
-  -- We define @tied@ as the fixpoint of the former to yield the former as a graph of type @IntMap.IntMap rep@, and apply the latter to said graph to yield the entry point, if any, from which to evaluate.
+  -- We define @tied@ as the fixpoint of the former to yield the former as a graph of type @IntMap.IntMap Term@, and apply the latter to said graph to yield the entry point, if any, from which to evaluate.
   let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
   pure (tied, ($ tied) <$> root)
 
-parseNode :: Syntax rep => A.Object -> A.Parser (IntMap.Key, IntMap.IntMap rep -> rep, Maybe (IntMap.IntMap rep -> rep))
+parseNode :: A.Object -> A.Parser (IntMap.Key, IntMap.IntMap Term -> Term, Maybe (IntMap.IntMap Term -> Term))
 parseNode o = do
-  edges <- o A..: pack "edges"
-  index <- o A..: pack "id"
-  let parseType attrs = \case
-        "string" -> const . string <$> attrs A..: pack "text"
-        "true"   -> pure (const (bool True))
-        "false"  -> pure (const (bool False))
-        "throw"  -> fmap throw <$> resolve (head edges)
-        "if"     -> liftA3 iff <$> findEdgeNamed "condition" <*> findEdgeNamed "consequence" <*> findEdgeNamed "alternative" <|> pure (const noop)
-        "block"  -> children
-        "module" -> children
-        t        -> A.parseFail ("unrecognized type: " <> t)
-      -- map the list of edges to a list of child nodes
-      children = fmap (foldr chain noop . zip [0..]) . sequenceA <$> traverse resolve edges
-      -- chain a statement before any following syntax by let-binding it. note that this implies call-by-value since any side effects in the statement must be performed before the let's body.
-      chain :: Syntax rep => (Int, rep) -> rep -> rep
-      chain (i, v) r = let_ (nameI i) v (const r)
-      resolve = resolveWith (const (pure ()))
-      resolveWith :: (A.Object -> A.Parser ()) -> A.Value -> A.Parser (IntMap.IntMap rep -> rep)
-      resolveWith f = A.withObject "edge" (\ edge -> do
-        sink <- edge A..: pack "sink"
-        attrs <- edge A..: pack "attrs"
-        f attrs
-        pure (IntMap.! sink))
-      findEdgeNamed :: (A.FromJSON a, Eq a) => a -> A.Parser (IntMap.IntMap rep -> rep)
-      findEdgeNamed name = foldMap (resolveWith (\ attrs -> attrs A..: pack "type" >>= guard . (== name))) edges
-  o A..: pack "attrs" >>= A.withObject "attrs" (\ attrs -> do
-    ty <- attrs A..: pack "type"
-    node <- parseType attrs ty
+  edges <- o A..: fromString "edges"
+  index <- o A..: fromString "id"
+  o A..: fromString "attrs" >>= A.withObject "attrs" (\ attrs -> do
+    ty <- attrs A..: fromString "type"
+    node <- parseType attrs edges ty
     pure (index, node, node <$ guard (ty == "module")))
+
+parseType :: A.Object -> [A.Value] -> String -> A.Parser (IntMap.IntMap Term -> Term)
```

Comment:
I pulled this and the rest of `parseNode`'s where clause to the top level, mostly to make it tidier and give them type signatures. The only real change otherwise is using `Term` constructors in `parseType` here instead of `Syntax` methods.